client/asset/eth/eth.go

```diff
@@ -593,101 +824,284 @@ func (c *coin) Value() uint64 {
+		w.log.Warnf("Swap gas estimate %d is greater than the server's configured value %d. Using live estimate + 10%.")
+		nSwap = gasEst * 11 / 10
+		if n == 1 && nSwap > oneSwap {
+			oneSwap = nSwap
+		}
+	}
+	return
+}
+
+// redeemGas gets an accurate estimate for redemption gas. We allow a DEX server
+// some latitude in adjusting the redemption gas, up to 2x our local estimate.
+func (w *AssetWallet) redeemGas(n int, nfo *dex.Asset) (oneGas, nGas uint64, err error) {
+	g := w.gases(nfo.Version)
+	if g == nil {
+		return 0, 0, fmt.Errorf("no gas table for redemption asset %d", nfo.ID)
+	}
+	redeemGas := nfo.RedeemSize
+	if g.Redeem > redeemGas {
+		// This probably won't happen.
+		w.log.Debugf("Server's reported redeem gas estimate %d is less than ours %d. Using our estimate.", g.Redeem, redeemGas)
+		redeemGas = g.Redeem
+	} else if nfo.RedeemSize > 2*g.Redeem {
+		return 0, 0, fmt.Errorf("server's reported redemption costs are more than twice the expected value. %d > 2 * %d",
+			nfo.RedeemSize, g.Redeem)
+	}
+	// Not concerned with the accuracy of nGas. It's never used outside of
+	// best case estimates.
+	return redeemGas, redeemGas + (uint64(n)-1)*g.RedeemAdd, nil
+}
+
+// approvalGas gets the best available estimate for an approval tx, which is
+// the greater of the asset's registered value and a live estimate. It is an
+// error if a live estimate cannot be retrieved, which will be the case if the
+// user's eth balance is insufficient to cover tx fees for the approval.
+func (w *AssetWallet) approvalGas(newGas *big.Int, cfg *dex.Asset) (uint64, error) {
+	ourGas := w.gases(cfg.Version)
+	if ourGas == nil {
+		return 0, fmt.Errorf("no gases known for %d version %d", w.assetID, cfg.Version)
+	}
+
+	approveGas := ourGas.Approve
+
+	if approveEst, err := w.estimateApproveGas(newGas); err != nil {
+		return 0, fmt.Errorf("error estimating approve gas: %v", err)
+	} else if approveEst > approveGas {
+		w.log.Warnf("Approve gas estimate %d is greater than the expected value %d. Using live estimate + 10%.")
+		return approveEst * 11 / 10, nil
+	}
+	return approveGas, nil
 }
 
 // ReturnCoins unlocks coins. This would be necessary in the case of a
 // canceled order.
-func (eth *ExchangeWallet) ReturnCoins(coins asset.Coins) error {
-	var amt uint64
+func (w *AssetWallet) ReturnCoins(coins asset.Coins) error {
+	var amt, fees uint64
 	for _, ci := range coins {
 		amt += ci.Value()
 		var addr common.Address
 		switch c := ci.(type) {
 		case *fundingCoin:
 			addr = c.addr
+		case *tokenFundingCoin:
+			fees += c.fees
+			addr = c.addr
 		default:
 			return fmt.Errorf("unknown coin type %T", c)
 		}
-		if addr != eth.addr {
-			return fmt.Errorf("coin is not funded by this wallet. coin address %s != our address %s", addr, eth.addr)
+		if addr != w.addr {
+			return fmt.Errorf("coin is not funded by this wallet. coin address %s != our address %s", addr, w.addr)
 		}
 	}
-
-	eth.lockedFunds.mtx.Lock()
-	defer eth.lockedFunds.mtx.Unlock()
```

Comment:
Still need to hold the lock to do `w.unlockFunds` I think.

Reply:
Whoops